File: app/src/test/java/com/gemmafit/video/TemporalMotionAnalyzerTest.java
package com.gemmafit.video;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.util.HashMap;
import java.util.Map;
import org.junit.Test;

public class TemporalMotionAnalyzerTest {
    @Test
    public void countsOneSquatRepFromTopBottomTopSequence() {
        TemporalMotionAnalyzer analyzer = new TemporalMotionAnalyzer();
        float[] angles = new float[] {175f, 160f, 130f, 100f, 85f, 105f, 135f, 170f};

        TemporalMotionAnalyzer.Result result = new TemporalMotionAnalyzer.Result();
        for (int i = 0; i < angles.length; i++) {
            result = analyzer.addSample(i, i * 200L, "squat", metrics("knee_angle", angles[i]));
        }

        assertEquals(1, result.getRepCount());
        assertEquals("top", result.getMovementPhase());
        assertNotNull(result.getCompletedRep());
        assertTrue(result.getRangeOfMotionDeg() >= 60f);
    }

    @Test
    public void rapidMovementRequiresConsecutiveSmoothedVelocityFrames() {
        TemporalMotionAnalyzer analyzer = new TemporalMotionAnalyzer();
        float[] angles = new float[] {180f, 120f, 30f, 120f};

        QualityFlag flag = null;
        for (int i = 0; i < angles.length; i++) {
            flag = analyzer
                .addSample(i, i * 100L, "squat", metrics("knee_angle", angles[i]))
                .getRapidFlag();
        }

        assertNotNull(flag);
        assertEquals(6, flag.getRule());
        assertEquals("rapid_movement", flag.getId());
        assertEquals(600f, flag.getThreshold(), 0.01f);
        assertTrue(flag.getValue() > 600f);
    }

    private static Map<String, Float> metrics(String key, float value) {
        Map<String, Float> metrics = new HashMap<>();
        metrics.put(key, value);
        return metrics;
    }
}

File: app/src/test/kotlin/com/gemmafit/video/TemporalMotionAnalyzerTest.kt
package com.gemmafit.video

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Test

class TemporalMotionAnalyzerBugfixTest {
    @Test
    fun analyze_usesFivePointSavitzkyGolaySmoothing() {
        val analyzer = TemporalMotionAnalyzer()
        val angles = listOf(10f, 20f, 30f, 40f, 50f)

        val result = angles.mapIndexed { index, angle ->
            analyzer.addSample(
                frameIndex = index,
                timestampMs = index * 100L,
                exercise = "squat",
                metrics = mapOf("knee_angle" to angle),
            )
        }.last()

        assertEquals(30f, result.primaryAngleDeg, 0.001f)
    }

    @Test
    fun analyze_sameTimestampDoesNotAmplifyVelocity() {
        val analyzer = TemporalMotionAnalyzer()
        analyzer.addSample(
            frameIndex = 0,
            timestampMs = 1_000L,
            exercise = "squat",
            metrics = mapOf("knee_angle" to 100f),
        )

        val result = analyzer.addSample(
            frameIndex = 1,
            timestampMs = 1_000L,
            exercise = "squat",
            metrics = mapOf("knee_angle" to 110f),
        )

        assertEquals(0f, result.smoothedVelocityDegS, 0.001f)
    }

    @Test
    fun completedRepProducesMotionFeatureWindow() {
        val analyzer = TemporalMotionAnalyzer()
        val angles = floatArrayOf(175f, 160f, 130f, 100f, 85f, 105f, 135f, 170f)
        var result = TemporalMotionAnalyzer.Result()

        for (i in angles.indices) {
            result = analyzer.addSample(
                frameIndex = i,
                timestampMs = i * 400L,
                exercise = "squat",
                metrics = mapOf("knee_angle" to angles[i]),
                confidenceFloor = 0.82f,
            )
        }

        val window = result.motionFeatureWindow
        assertNotNull(window)
        requireNotNull(window)
        assertEquals("motion_feature_window_v1", window.schemaVersion)
        assertEquals("REP_COMPLETED", window.trigger)
        assertEquals("squat", window.exercise)
        assertEquals("controlled", window.derivedLabels.tempoBand)
        assertEquals(true, window.derivedLabels.repCompleted)
        assertEquals(85f, window.features.kneeAngleMin ?: 0f, 0.001f)
        assertEquals(170f, window.features.kneeAngleMax ?: 0f, 0.001f)
        assertEquals(0.82f, window.features.confidenceFloor, 0.001f)
    }

    @Test
    fun bodyweightOrGobletSquatUsesKneeAngleForCompletedRep() {
        val analyzer = TemporalMotionAnalyzer()
        val angles = floatArrayOf(175f, 160f, 130f, 100f, 85f, 105f, 135f, 170f)
        var result = TemporalMotionAnalyzer.Result()

        for (i in angles.indices) {
            result = analyzer.addSample(
                frameIndex = i,
                timestampMs = i * 400L,
                exercise = "bodyweight_or_goblet_squat",
                metrics = mapOf("knee_angle" to angles[i], "hip_hinge" to 48f),
                confidenceFloor = 0.84f,
            )
        }

        val window = result.motionFeatureWindow
        assertNotNull(window)
        requireNotNull(window)
        assertEquals("bodyweight_or_goblet_squat", window.exercise)
        assertEquals("REP_COMPLETED", window.trigger)
        assertEquals(1, result.repCount)
        assertEquals(85f, window.features.kneeAngleMin ?: 0f, 0.001f)
    }
}
